server/src/main/java/org/candlepin/resource/validation/DTOValidator.java

```diff
@@ -0,0 +1,239 @@
+     */
+    @SafeVarargs
+    public final void validateMapElementsNotNull(Supplier<? extends Map<?, ?>> getter1,
+        Supplier<? extends Map<?, ?>> getter2, Supplier<? extends Map<?, ?>> getter3,
+        Supplier<? extends Map<?, ?>>... getters) {
+
+        validateMapElementsNotNull(getter1, getter2, getter3);
+
+        for (Supplier<? extends Map<?, ?>> getter : getters) {
+            Map<?, ?> map = getter.get();
+            if (isMapInvalid(map)) {
+                throw new IllegalArgumentException("map contains null elements");
+            }
+        }
+    }
+
+    /*
+     * ======================================================================================================
+     * The following methods are overloaded versions of validateMapElementsNotNull for one, two and
+     * three getters respectively, as a performance optimization to avoid the implicit creation of an array
+     * object that happens during the vararg method call. There are currently no DTOs that contain more than
+     * three maps that require null element validation, so this optimisation would cover 100% of cases
+     * for now, while allowing us to validate more than 3 maps at a time in the future.
+     * ======================================================================================================
+     */
+
+    /**
+     * Works the same as {@link #validateMapElementsNotNull(Supplier, Supplier, Supplier, Supplier...)}
+     */
+    public void validateMapElementsNotNull(Supplier<? extends Map<?, ?>> getter) {
+        Map<?, ?> map = getter.get();
+        if (isMapInvalid(map)) {
+            throw new IllegalArgumentException("map contains null elements");
+        }
+    }
+
+    /**
+     * Works the same as {@link #validateMapElementsNotNull(Supplier, Supplier, Supplier, Supplier...)}
+     */
+    public void validateMapElementsNotNull(Supplier<? extends Map<?, ?>> getter1,
+        Supplier<? extends Map<?, ?>> getter2) {
+        Map<?, ?> map1 = getter1.get();
+        Map<?, ?> map2 = getter2.get();
+        if (isMapInvalid(map1) ||
+            isMapInvalid(map2)) {
+            throw new IllegalArgumentException("map contains null elements");
+        }
+    }
+
+    /**
+     * Works the same as {@link #validateMapElementsNotNull(Supplier, Supplier, Supplier, Supplier...)}
+     */
+    public void validateMapElementsNotNull(Supplier<? extends Map<?, ?>> getter1,
+        Supplier<? extends Map<?, ?>> getter2, Supplier<? extends Map<?, ?>> getter3) {
+        Map<?, ?> map1 = getter1.get();
+        Map<?, ?> map2 = getter2.get();
+        Map<?, ?> map3 = getter3.get();
+        if (isMapInvalid(map1) ||
+            isMapInvalid(map2) ||
+            isMapInvalid(map3)) {
+            throw new IllegalArgumentException("map contains null elements");
+        }
+    }
+
+    private boolean isMapInvalid(Map<?, ?> map) {
+        return map != null &&
+            (map.values().stream().anyMatch(Objects::isNull) ||
```

Comment:
Map has methods `map.containsKey(null)` and `map.containsValue(null)`. We could use those.

Reply:
Apparently, some implementations of `Map` such as `ConcurrentHashMap` will throw an NPE if you pass null to `containsValue`, so I'll keep this as is